onnxruntime/core/framework/execution_frame.cc

```diff
@@ -831,7 +840,20 @@ AllocatorPtr ExecutionFrame::GetAllocatorImpl(const OrtDevice& info) const {
 // This method is not thread safe!
 // Return S_OK and nullptr if index map to a value that is an unused optional input/output
 Status ExecutionFrame::CreateNodeOutputMLValueImpl(OrtValue& ort_value, int ort_value_idx, const TensorShape* shape) {
+#ifndef ENABLE_TRAINING
   return AllocateAsPerAllocationPlan(ort_value, ort_value_idx, shape);
+#else
+  try {
+    auto status = AllocateAsPerAllocationPlan(ort_value, ort_value_idx, shape);
+    return status;
+  } catch (const std::exception& e) {
+    LOGS(session_state_.Logger(), WARNING)
+        << "Exception caught when allocating memory for ort_value with index: " << ort_value_idx
+        << "so clean up all_values_";
+    static_cast<void>(ReleaseAllMLValues());
```

Comment:
This look a bit weird. Maybe we can have a void return value for ReleaseAllMLValues if its logic is pretty straightforward.